node/core/approval-voting/src/aux_schema/mod.rs

```diff
@@ -0,0 +1,531 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Auxiliary DB schema, accessors, and writers for on-disk persisted approval storage
+//! data.
+//!
+//! We persist data to disk although it is not intended to be used across runs of the
+//! program. This is because under medium to long periods of finality stalling, for whatever
+//! reason that may be, the amount of data we'd need to keep would be potentially too large
+//! for memory.
+//!
+//! With tens or hundreds of parachains, hundreds of validators, and parablocks
+//! in every relay chain block, there can be a humongous amount of information to reference
+//! at any given time.
+//!
+//! As such, we provide a function from this module to clear the database on start-up.
+//! In the future, we may use a temporary DB which doesn't need to be wiped, but for the
+//! time being we share the same DB with the rest of Substrate.
+
+// TODO https://github.com/paritytech/polkadot/1975: remove this
+#![allow(unused)]
+
+use sc_client_api::backend::AuxStore;
+use polkadot_node_primitives::approval::{DelayTranche, RelayVRF};
+use polkadot_primitives::v1::{
+	ValidatorIndex, GroupIndex, CandidateReceipt, SessionIndex, CoreIndex,
+	BlockNumber, Hash, CandidateHash,
+};
+use sp_consensus_slots::SlotNumber;
+use parity_scale_codec::{Encode, Decode};
+
+use std::collections::{BTreeMap, HashMap};
+use std::collections::hash_map::Entry;
+use bitvec::{vec::BitVec, order::Lsb0 as BitOrderLsb0};
+
+use super::Tick;
+
+#[cfg(test)]
+mod tests;
+
+const STORED_BLOCKS_KEY: &[u8] = b"Approvals_StoredBlocks";
+
+/// Metadata regarding a specific tranche of assignments for a specific candidate.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct TrancheEntry {
+	tranche: DelayTranche,
+	// Assigned validators, and the instant we received their assignment, rounded
+	// to the nearest tick.
+	assignments: Vec<(ValidatorIndex, Tick)>,
+}
+
+/// Metadata regarding approval of a particular candidate within the context of some
+/// particular block.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct ApprovalEntry {
+	tranches: Vec<TrancheEntry>,
+	backing_group: GroupIndex,
+	// When the next wakeup for this entry should occur. This is either to
+	// check a no-show or to check if we need to broadcast an assignment.
+	next_wakeup: Tick,
+	our_assignment: Option<OurAssignment>,
+	// `n_validators` bits.
+	assignments: BitVec<BitOrderLsb0, u8>,
+	approved: bool,
+}
+
+/// Metadata regarding approval of a particular candidate.
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub(crate) struct CandidateEntry {
+	candidate: CandidateReceipt,
+	session: SessionIndex,
+	// Assignments are based on blocks, so we need to track assignments separately
+	// based on the block we are looking at.
+	block_assignments: BTreeMap<Hash, ApprovalEntry>,
```

Comment:
I suppose the `Hash` here is the hash of the relay chain block where inclusion happens?  It's ambiguious form this comment.

Reply:
Yes, that's right.

Reply:
Would it be worthwhile to define a `BlockHash` newtype adjacent to `CandidateHash`? It won't change anything re: type safety, but it might make it even harder to confuse what each hash represents.

Reply:
"worthwhile" is always a question of opportunity cost :) . I wouldn't object to such a change, but I don't think it's in the scope of this PR.

Reply:
Agree, that would be its own thing. I'll extract an issue.